# Fetch Rewards - Assessment

In [87]:
import pandas as pd
import json
from datetime import datetime

# Loading the json gzipped file into a dataframe

In [2]:
#loading the json gzipped file into a dataframe
receipts = pd.read_json(r"D:\receipts.json.gz", compression="gzip", lines=True)

In [3]:
receipts

_id  bonusPointsEarned  \
0     {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1     {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2     {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3     {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4     {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   
...                                    ...                ...   
1114  {'$oid': '603cc0630a720fde100003e6'}               25.0   
1115  {'$oid': '603d0b710a720fde1000042a'}                NaN   
1116  {'$oid': '603cf5290a720fde10000413'}                NaN   
1117  {'$oid': '603ce7100a7217c72c000405'}               25.0   
1118  {'$oid': '603c4fea0a7217c72c000389'}                NaN   

                                bonusPointsEarnedReason  \
0     Receipt number 2 completed, bonus point schedu...   
1     Receipt number 5 completed, bonus point schedu...   
2                            All-receipts receipt bonus   
3                            All-receipts receipt bonus   
4                            All-receipts receipt bonus   
...                                                 ...   
1114                        COMPLETE_NONPARTNER_RECEIPT   
1115                                                NaN   
1116                                                NaN   
1117                        COMPLETE_NONPARTNER_RECEIPT   
1118                                                NaN   

                    createDate               dateScanned  \
0     {'$date': 1609687531000}  {'$date': 1609687531000}   
1     {'$date': 1609687483000}  {'$date': 1609687483000}   
2     {'$date': 1609687537000}  {'$date': 1609687537000}   
3     {'$date': 1609687534000}  {'$date': 1609687534000}   
4     {'$date': 1609687506000}  {'$date': 1609687506000}   
...                        ...                       ...   
1114  {'$date': 1614594147000}  {'$date': 1614594147000}   
1115  {'$date': 1614613361873}  {'$date': 1614613361873}   
1116  {'$date': 1614607657664}  {'$date': 1614607657664}   
1117  {'$date': 1614604048000}  {'$date': 1614604048000}   
1118  {'$date': 1614565354962}  {'$date': 1614565354962}   

                  finishedDate                modifyDate  \
0     {'$date': 1609687531000}  {'$date': 1609687536000}   
1     {'$date': 1609687483000}  {'$date': 1609687488000}   
2                          NaN  {'$date': 1609687542000}   
3     {'$date': 1609687534000}  {'$date': 1609687539000}   
4     {'$date': 1609687511000}  {'$date': 1609687511000}   
...                        ...                       ...   
1114                       NaN  {'$date': 1614594148000}   
1115                       NaN  {'$date': 1614613361873}   
1116                       NaN  {'$date': 1614607657664}   
1117                       NaN  {'$date': 1614604049000}   
1118                       NaN  {'$date': 1614565354962}   

             pointsAwardedDate  pointsEarned              purchaseDate  \
0     {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1     {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                          NaN           5.0  {'$date': 1609632000000}   
3     {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4     {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   
...                        ...           ...                       ...   
1114                       NaN          25.0  {'$date': 1597622400000}   
1115                       NaN           NaN                       NaN   
1116                       NaN           NaN                       NaN   
1117                       NaN          25.0  {'$date': 1597622400000}   
1118                       NaN           NaN                       NaN   

      purchasedItemCount                             rewardsReceiptItemList  \
0                    5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                    2.0  [{'barcode': '4011', 'description'

**Converting all the Date columns to %Y-%m-%d %H:%M:%S format, making it more readable**

In [4]:
def date_converter(x):
    try:
        return(datetime.utcfromtimestamp(int(x['$date'])/1000).strftime('%Y-%m-%d %H:%M:%S'))
    except TypeError:
        return(None)


**Extracting values from json columns and getting the date in right format**

In [7]:
receipts['_id'] = receipts['_id'].apply(lambda x: x['$oid'])
receipts['createDate'] = receipts['createDate'].apply(lambda x: date_converter(x))
receipts['dateScanned'] = receipts['dateScanned'].apply(lambda x: date_converter(x))
receipts['finishedDate'] = receipts['finishedDate'].apply(lambda x: date_converter(x))
receipts['modifyDate'] = receipts['modifyDate'].apply(lambda x: date_converter(x))
receipts['pointsAwardedDate'] = receipts['pointsAwardedDate'].apply(lambda x: date_converter(x))
receipts['purchaseDate'] = receipts['purchaseDate'].apply(lambda x: date_converter(x))

## Total Spent is missing for 435 rows of data

In [9]:
receipts["totalSpent"].isna().sum()

435

In [24]:
receipts["pointsEarned"].isna().sum()

510

## Over 500 pointsEarned values are malformed or null - this could break analytics unless cleaned.

In [72]:
malformed_values = receipts["pointsEarned"][~receipts["pointsEarned"].astype(str).str.replace('.', '', 1).str.isnumeric()]
print(f"Malformed 'pointsEarned' values: {len(malformed_values)}")

Malformed 'pointsEarned' values: 510


## All the id in the receipt dataframe are unique

In [30]:
receipts['_id'].nunique()

1119

## purchaseDate column are formatted, but they still have many null values

In [32]:
receipts.purchaseDate.isna().sum()

448

# Receipts Data Findings:

- Nearly 40% of receipts are missing spend/purchase/item info, indicating potentially incomplete ingestion or partial scans.
- Over 500 pointsEarned values are malformed (non-numeric or strings like "N/A", etc.) — this could break analytics unless cleaned.
- Date columns are present and now consistently formatted, but many are null.

-----------------------------------------------------------------------------------

In [38]:
receipts['rewardsReceiptItemList'].str.len().sum()

6941.0

## rewardsReceiptItemList has a lenght of 6941. Exploding the column into a separate 'items_df' dataframe, we get:

In [35]:
items_data = []

for idx, row in receipts.iterrows():
    items = row["rewardsReceiptItemList"]
    if isinstance(items, list):
        for item in items:
            item["receipt_id"] = row["_id"]
            item["userId"] = row["userId"]
            items_data.append(item)

# Convert the list of dictionaries into a DataFrame
items_df = pd.DataFrame(items_data)

In [37]:
items_df

barcode                                        description  \
0             4011                                     ITEM NOT FOUND   
1             4011                                     ITEM NOT FOUND   
2     028400642255  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...   
3              NaN                                                NaN   
4             4011                                     ITEM NOT FOUND   
...            ...                                                ...   
6936    B07BRRLSVC  thindust summer face mask - sun protection nec...   
6937    B076FJ92M4  mueller austria hypergrind precision electric ...   
6938    B07BRRLSVC  thindust summer face mask - sun protection nec...   
6939    B076FJ92M4  mueller austria hypergrind precision electric ...   
6940    B07BRRLSVC  thindust summer face mask - sun protection nec...   

     finalPrice itemPrice needsFetchReview partnerItemId  \
0         26.00     26.00            False             1   
1             1         1              NaN             1   
2         10.00     10.00             True             2   
3           NaN       NaN            False             1   
4         28.00     28.00            False             1   
...         ...       ...              ...           ...   
6936      11.99     11.99              NaN             1   
6937      22.97     22.97              NaN             0   
6938      11.99     11.99              NaN             1   
6939      22.97     22.97              NaN             0   
6940      11.99     11.99              NaN             1   

     preventTargetGapPoints  quantityPurchased userFlaggedBarcode  \
0                      True                5.0               4011   
1                       NaN                1.0                NaN   
2                      True                1.0       028400642255   
3                      True                NaN               4011   
4                      True                4.0               4011   
...                     ...                ...                ...   
6936                    NaN                1.0                NaN   
6937                    NaN                1.0                NaN   
6938                    NaN                1.0                NaN   
6939                    NaN                1.0                NaN   
6940                    NaN                1.0                NaN   

     userFlaggedNewItem  ... itemNumber  originalMetaBriteQuantityPurchased  \
0                  True  ...        NaN                                 NaN   
1                   NaN  ...        NaN                                 NaN   
2                  True  ...        NaN                                 NaN   
3                  True  ...        NaN                                 NaN   
4                  True  ...        NaN                                 NaN   
...                 ...  ...        ...                                 ...   
6936                NaN  ...        NaN                                 NaN   
6937                NaN  ...        NaN                                 NaN   
6938                NaN  ...        NaN                                 NaN   
6939                NaN  ...        NaN                                 NaN   
6940                NaN  ...        NaN                                 NaN   

     pointsEarned targetPrice competitiveProduct originalFinalPrice  \
0             NaN         NaN                NaN                NaN   
1             NaN         NaN                NaN                NaN   
2             NaN         NaN                NaN                NaN   
3             NaN         NaN                NaN                NaN   
4             NaN         NaN                NaN                NaN   
...           ...         ...                ...                ...   
6936          NaN         NaN                NaN                NaN   
6937          NaN         NaN                NaN                NaN

In [90]:
items_df.columns

Index(['barcode', 'description', 'finalPrice', 'itemPrice', 'needsFetchReview',
       'partnerItemId', 'preventTargetGapPoints', 'quantityPurchased',
       'userFlaggedBarcode', 'userFlaggedNewItem', 'userFlaggedPrice',
       'userFlaggedQuantity', 'receipt_id', 'userId', 'needsFetchReviewReason',
       'pointsNotAwardedReason', 'pointsPayerId', 'rewardsGroup',
       'rewardsProductPartnerId', 'userFlaggedDescription',
       'originalMetaBriteBarcode', 'originalMetaBriteDescription', 'brandCode',
       'competitorRewardsGroup', 'discountedItemPrice',
       'originalReceiptItemText', 'itemNumber',
       'originalMetaBriteQuantityPurchased', 'pointsEarned', 'targetPrice',
       'competitiveProduct', 'originalFinalPrice',
       'originalMetaBriteItemPrice', 'deleted', 'priceAfterCoupon',
       'metabriteCampaignId', 'quantityPurchased_str', 'itemPrice_str',
       'finalPrice_str', 'quantityPurchased_num', 'itemPrice_num',
       'finalPrice_num'],
      dtype='object')

## Nearly 50% of the barcodes in the dataframe are null

In [39]:
items_df.barcode.isna().sum()

3851

## Missing Description: Product label not captured, which limits usability

In [91]:
items_df.description.isna().sum()

381

In [41]:
items_df.quantityPurchased.value_counts()

1.0     5628
2.0      622
4.0      170
3.0      134
5.0      101
6.0       37
8.0       22
10.0      15
9.0       13
7.0       13
12.0       6
17.0       3
14.0       3
Name: quantityPurchased, dtype: int64

# Loading the Brands Data

In [52]:
brands = pd.read_json(r"D:\brands.json.gz", compression="gzip", lines=True)

In [65]:
brands

_id       barcode            category  \
0     {'$oid': '601ac115be37ce2ead437551'}  511111019862              Baking   
1     {'$oid': '601c5460be37ce2ead43755f'}  511111519928           Beverages   
2     {'$oid': '601ac142be37ce2ead43755d'}  511111819905              Baking   
3     {'$oid': '601ac142be37ce2ead43755a'}  511111519874              Baking   
4     {'$oid': '601ac142be37ce2ead43755e'}  511111319917      Candy & Sweets   
...                                    ...           ...                 ...   
1162  {'$oid': '5f77274dbe37ce6b592e90c0'}  511111116752              Baking   
1163  {'$oid': '5dc1fca91dda2c0ad7da64ae'}  511111706328  Breakfast & Cereal   
1164  {'$oid': '5f494c6e04db711dd8fe87e7'}  511111416173      Candy & Sweets   
1165  {'$oid': '5a021611e4b00efe02b02a57'}  511111400608             Grocery   
1166  {'$oid': '6026d757be37ce6369301468'}  511111019930              Baking   

          categoryCode                                                cpg  \
0               BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1            BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2               BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3               BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4     CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   
...                ...                                                ...   
1162            BAKING  {'$ref': 'Cogs', '$id': {'$oid': '5f77274dbe37...   
1163               NaN  {'$ref': 'Cogs', '$id': {'$oid': '53e10d6368ab...   
1164  CANDY_AND_SWEETS  {'$ref': 'Cogs', '$id': {'$oid': '5332fa12e4b0...   
1165               NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5f6e4b0...   
1166            BAKING  {'$id': {'$oid': '6026d757be37ce6369301467'}, ...   

                           name  topBrand                      brandCode  
0     test brand @1612366101024       0.0                            NaN  
1                     Starbucks       0.0                      STARBUCKS  
2     test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3     test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4     test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827  
...                         ...       ...                            ...  
1162  test brand @1601644365844       NaN                            NaN  
1163        Dippin Dots® Cereal       NaN             DIPPIN DOTS CEREAL  
1164  test brand @1598639215217       NaN  TEST BRANDCODE @1598639215217  
1165          LIPTON TEA Leaves       0.0              LIPTON TEA Leaves  
1166  test brand @1613158231643       0.0  TEST BRANDCODE @1613158231644  

[1167 rows x 8 columns]

## cpg column, which is a key value pair into 2 different columns

In [53]:
brands_json = pd.json_normalize(brands['cpg'])

In [54]:
brands_json

$ref                  $id.$oid
0     Cogs  601ac114be37ce2ead437550
1     Cogs  5332f5fbe4b03c9a25efd0ba
2     Cogs  601ac142be37ce2ead437559
3     Cogs  601ac142be37ce2ead437559
4     Cogs  5332fa12e4b03c9a25efd1e7
...    ...                       ...
1162  Cogs  5f77274dbe37ce6b592e90bf
1163  Cogs  53e10d6368abd3c7065097cc
1164  Cogs  5332fa12e4b03c9a25efd1e7
1165  Cogs  5332f5f6e4b03c9a25efd0b4
1166  Cogs  6026d757be37ce6369301467

[1167 rows x 2 columns]

#### $ref value counts below - 'Cogs' make up more than 90%

In [62]:
brands_json['$ref'].value_counts()

Cogs    1020
Cpgs     147
Name: $ref, dtype: int64

In [68]:
brands['_id'] = brands['_id'].apply(lambda x: x['$oid'])
brands

_id       barcode            category  \
0     601ac115be37ce2ead437551  511111019862              Baking   
1     601c5460be37ce2ead43755f  511111519928           Beverages   
2     601ac142be37ce2ead43755d  511111819905              Baking   
3     601ac142be37ce2ead43755a  511111519874              Baking   
4     601ac142be37ce2ead43755e  511111319917      Candy & Sweets   
...                        ...           ...                 ...   
1162  5f77274dbe37ce6b592e90c0  511111116752              Baking   
1163  5dc1fca91dda2c0ad7da64ae  511111706328  Breakfast & Cereal   
1164  5f494c6e04db711dd8fe87e7  511111416173      Candy & Sweets   
1165  5a021611e4b00efe02b02a57  511111400608             Grocery   
1166  6026d757be37ce6369301468  511111019930              Baking   

          categoryCode                                                cpg  \
0               BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1            BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2               BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3               BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4     CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   
...                ...                                                ...   
1162            BAKING  {'$ref': 'Cogs', '$id': {'$oid': '5f77274dbe37...   
1163               NaN  {'$ref': 'Cogs', '$id': {'$oid': '53e10d6368ab...   
1164  CANDY_AND_SWEETS  {'$ref': 'Cogs', '$id': {'$oid': '5332fa12e4b0...   
1165               NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5f6e4b0...   
1166            BAKING  {'$id': {'$oid': '6026d757be37ce6369301467'}, ...   

                           name  topBrand                      brandCode  
0     test brand @1612366101024       0.0                            NaN  
1                     Starbucks       0.0                      STARBUCKS  
2     test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3     test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4     test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827  
...                         ...       ...                            ...  
1162  test brand @1601644365844       NaN                            NaN  
1163        Dippin Dots® Cereal       NaN             DIPPIN DOTS CEREAL  
1164  test brand @1598639215217       NaN  TEST BRANDCODE @1598639215217  
1165          LIPTON TEA Leaves       0.0              LIPTON TEA Leaves  
1166  test brand @1613158231643       0.0  TEST BRANDCODE @1613158231644  

[1167 rows x 8 columns]

### All the brand ids are unique

In [71]:
brands._id.nunique()

1167

## Brand Code has lot of rows which has null value

In [63]:
brands['brandCode'].value_counts()

                                   35
HUGGIES                             2
GOODNITES                           2
KELSEN                              1
THE GLENLIVET CARIBBEAN RESERVE     1
                                   ..
TEST BRANDCODE @1598711015496       1
ABSOLUT® MANDRIN                    1
TEST BRANDCODE @1598711015353       1
TEST BRANDCODE @1610660035741       1
F WHITLOCK AND SONS                 1
Name: brandCode, Length: 897, dtype: int64

In [64]:
brands['brandCode'].isna().sum()

234

### Around 10% are missing category values, which may impact category-based reporting.

In [67]:
brands.category.isna().sum()

155

In [94]:
brands[brands['topBrand'] == 1.0]

_id       barcode             category  \
58    5c76d3cd95144c5375687b4f  511111106876              Grocery   
109   585a9645e4b03e62d1ce0e79  511111801757               Snacks   
115   5887a372e4b02187f85cdad9  511111001119               Snacks   
116   57ed0697e4b072ac2294b8f2  511111101895  Condiments & Sauces   
152   5c45f91b87ff3552f950f027  511111204923              Grocery   
192   585a963ce4b03e62d1ce0e78  511111001768               Snacks   
245   5d66961cee7f2d201c7281cc  511111812449            Magazines   
258   57c0827de4b0718ff5fcb037  511111902461                 Baby   
271   585a9637e4b03e62d1ce0e77  511111501770   Breakfast & Cereal   
278   585a967fe4b03e62d1ce0e80  511111801689               Snacks   
344   5a4d1f06e4b0d5c3fd86b678  511111700531               Snacks   
428   5c9a975495144c182bbc7270  511111208532              Grocery   
467   5c409ab4cd244a3539b84162  511111004790               Baking   
471   585a966be4b03e62d1ce0e7e  511111901709               Snacks   
580   5c76dd4d95144c5375687b52  511111806899                 Baby   
581   585a964ce4b03e62d1ce0e7a  511111301745               Snacks   
721   585a965be4b03e62d1ce0e7c  511111601722               Snacks   
723   55b6309ce4b0d8e685c1421b  511111702665            Beverages   
773   585a9618e4b03e62d1ce0e75  511111201793               Snacks   
848   585a961fe4b03e62d1ce0e76  511111701781               Snacks   
943   5c6d51d695144c607914426f  511111806509   Breakfast & Cereal   
978   5db3288aee7f2d6de4248977  511111312949                 Baby   
1012  5c4637ba87ff35681e840d57  511111605058                Dairy   
1018  585a9686e4b03e62d1ce0e81  511111601678               Snacks   
1020  592486bde410d61fcea3d121  511111300939               Frozen   
1023  5c6c2f9295144c6db8fe1977  511111106432            Beverages   
1030  592486bde410d61fcea3d122  511111800927               Frozen   
1036  5db32879ee7f2d6de4248976  511111112938                 Baby   
1061  5a4d1ec9e4b0bcb2c74ea77d  511111900542               Snacks   
1074  5c7d9cb395144c337a3cbfbb  511111707202                 Baby   
1109  585a9661e4b03e62d1ce0e7d  511111401711               Snacks   

     categoryCode                                                cpg  \
58            NaN  {'$ref': 'Cogs', '$id': {'$oid': '5c76d2059514...   
109           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
115           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
116           NaN  {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...   
152           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5c45f8b087ff...   
192           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
245           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5c4f20b01b41...   
258           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f7a7e4b0...   
271           NaN  {'$ref': 'Cogs', '$id': {'$oid': '53e10d6368ab...   
278           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
344           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
428           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5c7470c89514...   
467           NaN  {'$ref': 'Cogs', '$id': {'$oid': '55b62995e4b0...   
471           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
580           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5c76dd109514...   
581           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
721           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
723           NaN  {'$ref': 'Cogs', '$id': {'$oid': '55b62995e4b0...   
773           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
848           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
943           NaN  {'$ref': 'Cogs', '$id': {'$oid': '5c6d51679514...   
978           NaN  {'$ref': 'Cogs', '$id': {'$oid': '550b2565e4b0...   
1012          NaN  {'$ref': 'Cogs', '$id': {'$oid': '5c45f8b087ff...   
1018          NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
1020       FROZEN 

## Unique number of brand code differ in the items and brands dataframe

In [92]:
items_df.brandCode.nunique()

227

In [93]:
brands.brandCode.nunique()

897

# Brand data findings:
- Brand table is relatively clean — all records have valid IDs and names.
- About 20% are missing brand codes, which may impact brand-related analysis
- About 13% are missing category values, which may impact category-based reporting.


# Loading the Users Data frame

In [77]:
users = pd.read_json(r"D:\users.json",lines=True)
users

_id  active               createdDate  \
0    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3    {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
..                                    ...     ...                       ...   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSource state  
0    {'$date': 1609687537858}     consumer        Email    WI  
1    {'$date': 1609687537858}     consumer        Email    WI  
2    {'$date': 1609687537858}     consumer        Email    WI  
3    {'$date': 1609687530597}     consumer        Email    WI  
4    {'$date': 1609687537858}     consumer        Email    WI  
..                        ...          ...          ...   ...  
490  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
491  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
492  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
493  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
494  {'$date': 1614963143204}  fetch-staff          NaN   NaN  

[495 rows x 7 columns]

## Cleaning the json _id columns, and the date columns

In [78]:
users['_id'] = users['_id'].apply(lambda x: x['$oid'])
users['createdDate'] = users['createdDate'].apply(lambda x: date_converter(x))
users['lastLogin'] = users['lastLogin'].apply(lambda x: date_converter(x))

In [79]:
users

_id  active          createdDate  \
0    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
1    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
2    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
3    5ff1e1eacfcf6c399c274ae6    True  2021-01-03 15:25:30   
4    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
..                        ...     ...                  ...   
490  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
491  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
492  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
493  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
494  54943462e4b07e684157a532    True  2014-12-19 14:21:22   

               lastLogin         role signUpSource state  
0    2021-01-03 15:25:37     consumer        Email    WI  
1    2021-01-03 15:25:37     consumer        Email    WI  
2    2021-01-03 15:25:37     consumer        Email    WI  
3    2021-01-03 15:25:30     consumer        Email    WI  
4    2021-01-03 15:25:37     consumer        Email    WI  
..                   ...          ...          ...   ...  
490  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
491  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
492  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
493  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
494  2021-03-05 16:52:23  fetch-staff          NaN   NaN  

[495 rows x 7 columns]

## High number of duplicate user records (over 50% are duplicates).

In [80]:
users._id.nunique()

212

## There is a discrepancy in the count of userId
- Unique userId in the receipts dataframe : 258
- Unique Id in the users dataframe : 212

## This may be due to missing userIds in the Users dataframe

In [86]:
receipts.userId.nunique()

258

## About 10% of the states listed are null

In [83]:
users.state.isna().sum()

56

## 62 lastLogin values are null, which will affect analysis on when the user was last using the account

In [85]:
users.lastLogin.isna().sum()

62

# Users Findings:
- High number of duplicate user records (over 50% are duplicates).
- Some users have invalid or missing state entries — possibly dirty manual entries or system errors.
- lastLogin is missing for several users — may indicate inactive or incomplete accounts.